pallets/anonymity-mining-claims/src/lib.rs

```diff
@@ -240,97 +310,269 @@ pub mod pallet {
+		let root_index = Self::next_unspent_root_index(unspent_resource_id);
+		println!("SUBSTRATE: root_index: {root_index:?}");
+		ensure!(
+			Self::get_unspent_roots(unspent_resource_id).len() >= 1,
+			Error::<T, I>::InvalidUnspentChainIds
+		);
 		// Update unspent root
-		CachedUnspentRoots::<T, I>::insert(unspent_tree_id, unspent_root_index, unspent_root);
+		CachedUnspentRoots::<T, I>::insert(unspent_resource_id, root_index, unspent_root);
+
+		NextUnspentRootIndex::<T, I>::mutate(unspent_resource_id, |i| {
+			*i = i.saturating_add(One::one()) % T::UnspentRootHistorySize::get()
+		});
+		Ok(())
 	}
 
 	/// Update spent root
 	fn update_spent_root(
-		spent_tree_id: T::TreeId,
-		spent_root_index: T::RootIndex,
+		spent_resource_id: ResourceId,
 		spent_root: T::Element,
-	) {
+	) -> Result<(), DispatchError> {
+		let root_index = Self::next_spent_root_index(spent_resource_id);
+		ensure!(
+			Self::get_spent_roots(spent_resource_id).len() >= 1,
+			Error::<T, I>::InvalidSpentChainIds
+		);
+
 		// Update spent root
-		CachedSpentRoots::<T, I>::insert(spent_tree_id, spent_root_index, spent_root);
-	}
+		CachedSpentRoots::<T, I>::insert(spent_resource_id, root_index, spent_root);
 
-	/// Handle claiming of AP tokens
-	pub fn claim_ap(
-		unspent_resource_id: ResourceId,
-		spent_resource_id: ResourceId,
-		recipient: T::AccountId,
-		amount: BalanceOf<T, I>,
-		merkle_root: T::Element,
-		latest_leaf_index: T::LeafIndex,
-		proof_data: ProofData<T::Element>,
-		reward_proof_data: RewardProofData<T::Element>,
-		unspent_root: T::Element,
-		spent_root: T::Element,
-		reward_nullifier_hash: T::Element,
-	) -> DispatchResultWithPostInfo {
-		// Handle proof verification
-		Self::handle_proof_verification(&proof_data)?;
-
-		let unspent_tree_id: T::TreeId = match unspent_resource_id.target_system() {
-			TargetSystem::Substrate(system) => system.tree_id.into(),
-			_ => {
-				ensure!(false, Error::<T, I>::InvalidResourceId);
-				T::TreeId::default()
-			},
-		};
-
-		let spent_tree_id: T::TreeId = match spent_resource_id.target_system() {
-			TargetSystem::Substrate(system) => system.tree_id.into(),
-			_ => {
-				ensure!(false, Error::<T, I>::InvalidResourceId);
-				T::TreeId::default()
-			},
-		};
-
-		let unspent_root_index = Self::next_unspent_root_index();
-		let spent_root_index = Self::next_spent_root_index();
-
-		NextUnspentRootIndex::<T, I>::mutate(|i| {
-			*i = i.saturating_add(One::one()) % T::UnspentRootHistorySize::get()
-		});
-		NextSpentRootIndex::<T, I>::mutate(|i| {
+		NextSpentRootIndex::<T, I>::mutate(spent_resource_id, |i| {
 			*i = i.saturating_add(One::one()) % T::UnspentRootHistorySize::get()
 		});
 
-		// Add reward_nullifier_hash
-		Self::add_reward_nullifier_hash(reward_nullifier_hash);
+		Ok(())
+	}
 
-		// Add nullifier on VAnchor
-		for nullifier in &proof_data.input_nullifiers {
-			T::VAnchor::add_nullifier_hash(Self::ap_vanchor_tree_id(), *nullifier)?;
+	/// Ensure valid unspent roots
```

Comment:
Need better function docs, will help speed up review process to understand the expectation of each function.